### ISSUE TITLE
[QUESTION] "overriding recipe for target" warnings in HEMCO 3.0.0 build

### ISSUE THREAD

**Issue by yantosca** [geoschem/HEMCO#87]
# Ask a question about HEMCO

I am developing in a branch off of GEOS-Chem 13.1.1 , which uses HEMCO 3.0.0.  I noticed a bunch of these warning messages:
```
src/HEMCO/src/Core/CMakeFiles/HCO.dir/depend.make:387: warning: overriding recipe for target `src/HEMCO/src/Core/CMakeFiles/HCO.dir/hcoio_read_mod.mod.stamp'
src/HEMCO/src/Core/CMakeFiles/HCO.dir/depend.make:363: warning: ignoring old recipe for target `src/HEMCO/src/Core/CMakeFiles/HCO.dir/hcoio_read_mod.mod.stamp'
src/HEMCO/src/Core/CMakeFiles/HCO.dir/depend.make:442: warning: overriding recipe for target `src/HEMCO/src/Core/CMakeFiles/HCO.dir/hcoio_write_mod.mod.stamp'
src/HEMCO/src/Core/CMakeFiles/HCO.dir/depend.make:421: warning: ignoring old recipe for target `src/HEMCO/src/Core/CMakeFiles/HCO.dir/hcoio_write_mod.mod.stamp'
src/HEMCO/src/Core/CMakeFiles/HCO.dir/depend.make:387: warning: overriding recipe for target `src/HEMCO/src/Core/CMakeFiles/HCO.dir/hcoio_read_mod.mod.stamp'
src/HEMCO/src/Core/CMakeFiles/HCO.dir/depend.make:363: warning: ignoring old recipe for target `src/HEMCO/src/Core/CMakeFiles/HCO.dir/hcoio_read_mod.mod.stamp'
src/HEMCO/src/Core/CMakeFiles/HCO.dir/depend.make:442: warning: overriding recipe for target `src/HEMCO/src/Core/CMakeFiles/HCO.dir/hcoio_write_mod.mod.stamp'
src/HEMCO/src/Core/CMakeFiles/HCO.dir/depend.make:421: warning: ignoring old recipe for target `src/HEMCO/src/Core/CMakeFiles/HCO.dir/hcoio_write_mod.mod.stamp'
```
Not sure what they mean.  The code builds and runs but maybe there is some leftover stuff in the HEMCO Makefiles?

**Comment by lizziel** [comment]
This warning is described in the [GNU make docs ](https://www.gnu.org/software/make/manual/html_node/Error-Messages.html)as this:

```
warning: overriding recipe for target `xxx'
warning: ignoring old recipe for target `xxx'
```
> GNU make allows only one recipe to be specified per target (except for double-colon rules). If you give a recipe for a target which already has been defined to have one, this warning is issued and the second recipe will overwrite the first. See Multiple Rules for One Target.

Not sure what the solution is but this is a start.

**Comment by jimmielin** [comment]
I think one of my changes at https://github.com/geoschem/HEMCO/pull/80 made this happen. I am not familiar with CMake enough to know what is tripping this, but the change in the PR#80 changed both files `hcoio_read_std_mod.F90` and `hcoio_read_mapl_mod.F90` to have the module name `HCOIO_Read_Mod`, and one of the files will be empty depending on pre-processor constants and which interface we are building for. This is intended so that future HCOIO interfaces can be added with the same format, and no changes to other HEMCO files are needed to make use of these, only the pre-processor switches need to be changed.

Perhaps this could be implemented in a better way that doesn't throw off warnings, but I am not knowledgeable enough with CMake. @LiamBindle Liam, may I ask if you have a better solution to this problem? A more detailed description of what we want to accomplish is in PR#80.

**Comment by LiamBindle** [comment]
I believe the issues is that these lines
https://github.com/geoschem/HEMCO/blob/051d59b391b2cad24c3ef1da96f54ad5a184ac78/src/Core/CMakeLists.txt#L31-L32
both declare a target for `hcoio_read_mod.mod`. In practice, the preprocessor blocks mean that only one `hcoio_read_mod.mod` exists after both files are compiled (so it works), but I don't think the build system understands what's happening here. 

<details>
  <summary>Some details</summary>
  The tricky nuance here is that preprocessor directives are intended for the compiler, not the build system. I don't think the build system really understands....if you think about it, it's the build system that specifies the PP definitions to the compilers.
</details>

I believe the canonical way to achieve the desired behaviour would be to **only declare the correct `hcoio_read_mod` as a dependency of `HCO`**. Rather than this

```cmake
add_library(HCO STATIC EXCLUDE_FROM_ALL
    ...
    hcoio_read_mapl_mod.F90  # declares hcoio_read_mod.mod target as depdency of HCO target
    hcoio_read_std_mod.F90   # declares hcoio_read_mod.mod (duplicate) target as depdency of HCO target
    hcoio_write_mapl_mod.F90 # declares hcoio_write_mod.mod target as depdency of HCO target
    hcoio_write_std_mod.F90  # declares hcoio_write_mod.mod (duplicate) target as depdency of HCO target
    ...
)
```
you could try this
```cmake
add_library(HCO STATIC EXCLUDE_FROM_ALL
    ...
    $<$<BOOL:${MAPL_ESMF}>: hcoio_read_mapl_mod.F90 hcoio_write_mapl_mod.F90>      # only incl. iff(MAPL_ESMF)
    $<$<NOT:$<BOOL:${MAPL_ESMF}>>: hcoio_read_std_mod.F90 hcoio_write_std_mod.F90> # only incl. iff(!MAPL_ESMF)
    ...
)
```

that way `hcoio_read_mapl_mod.F90` and `hcoio_write_mapl_mod.F90` are only included in the build if `${MAPL_ESMF}` is true, otherwise `hcoio_read_std_mod.F90` and `hcoio_write_std_mod.F90` are included. (The syntax is ugly but generator expressions do work well and most importantly the semantics are correct).

**References**
* [CMake generator expressions](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html)

**Comment by jimmielin** [comment]
Hi Liam, thank you so much for the detailed writeup. If I understand it correctly, then we could potentially declare a cache variable `HEMCO_INTERFACE` then we could allow switching for different HEMCO I/O modules, e.g.,

```cmake
add_library(HCO STATIC EXCLUDE_FROM_ALL
    ...
    $<STREQUAL:${HEMCO_INTERFACE},"classic">: hcoio_read_std_mod.F90 hcoio_write_std_mod.F90>
    $<STREQUAL:${HEMCO_INTERFACE},"mapl">: hcoio_read_mapl_mod.F90 hcoio_write_mapl_mod.F90>
    $<STREQUAL:${HEMCO_INTERFACE},"future">: hcoio_read_future_mod.F90 hcoio_write_future_mod.F90>
    ...
)
```

This would be perfect and we could just declare a different `HEMCO_INTERFACE` variable in CMake, then use that to control (1) the module file being compiled, and (2) the pre-processor constant. So there wouldn't be the inverted hierarchy of having pre-processor constants telling the make process what to do.

If this sounds good I can look into fixing the CMake files, but we also need to document this behavior about the `HEMCO_INTERFACE` cache variable so other HEMCO developers don't get tripped up compiling the wrong I/O module. But having a flexible switch will definitely be more future proof as more models are coupled to HEMCO.

**Comment by LiamBindle** [comment]
Exactly! IMO that's be the best solution. That way the behaviour is explicit and well defined according to the `HEMCO_INTERFACE` variable. That's a mechanism super projects can use to control the build in a well defined manner.

Minor correction: I think you're missing a leading `$<` in those generator expressions above. I think it should be 
```cmake
add_library(HCO STATIC EXCLUDE_FROM_ALL
    ...
    $<$<STREQUAL:${HEMCO_INTERFACE},"classic">: hcoio_read_std_mod.F90 hcoio_write_std_mod.F90>
    $<$<STREQUAL:${HEMCO_INTERFACE},"mapl">: hcoio_read_mapl_mod.F90 hcoio_write_mapl_mod.F90>
    $<$<STREQUAL:${HEMCO_INTERFACE},"future">: hcoio_read_future_mod.F90 hcoio_write_future_mod.F90>
    ...
)
```
(The synax is kind of awful)

**Comment by jimmielin** [comment]
I was working on this earlier and it seems like I couldn't get the generator expression working - I could get `${HEMCO_INTERFACE}` in `src/Core` fine but after the generator runs, the source files for `hcoio_...` weren't added to the final makefiles.

However, this solution worked for me: I first created a list of source files that wouldn't vary according to `HEMCO_INTERFACE` and then appended to it using conditionals:

```cmake
set(src_HCOCore
	hco_arr_mod.F90
	hco_calc_mod.F90
	...
)

if(HEMCO_INTERFACE STREQUAL standalone)
	list(APPEND src_HCOCore hcoio_read_std_mod.F90 hcoio_write_std_mod.F90)
elseif(HEMCO_INTERFACE STREQUAL gcclassic)
	list(APPEND src_HCOCore hcoio_read_std_mod.F90 hcoio_write_std_mod.F90)
elseif(HEMCO_INTERFACE STREQUAL mapl)
	list(APPEND src_HCOCore hcoio_read_mapl_mod.F90 hcoio_write_mapl_mod.F90)
elseif(HEMCO_INTERFACE STREQUAL future)
	list(APPEND src_HCOCore hcoio_read_future_mod.F90 hcoio_write_future_mod.F90)
endif()

add_library(HCO STATIC EXCLUDE_FROM_ALL
	${src_HCOCore}
)
```

This approach works fine. I'll push this to a branch and PR later once I test it works with the other interfaces. The issue is that unfortunately CESM doesn't have CMake for now (nor does WRF-GC) so the pre-processor constants still need to be there for the foreseeable future. But at least it fixes the issue with the warning messages!

**Comment by jimmielin** [comment]
Hi all, I PRed a fix for this issue in here: https://github.com/geoschem/HEMCO/pull/90

It might be more helpful to continue this discussion here if needed since there is more context. I've tested this with standalone, GC-Classic and GCHP. I think it should also support GEOS fine (I edited CMakeLists.txt accordingly) but it would be helpful to give them all a test to make sure there is no regression.

WRF-GC and CESM2-GC don't use CMake so they should behave as before, but this means that the C-preprocessor variables are here to stay...

Best,
Haipeng
